### DISCUSSION TITLE
Getting Cursorless into other IDEs/apps

### DISCUSSION THREAD

**Discussion by pokey** [cursorless-dev/cursorless#2189] [General]
We recently decided that the approach to integrating Cursorless into other IDEs need not be a one-size-fits-all approach. There are three possible approaches:

- Bundling Cursorless into a language server, and letting the IDE implement a language client that talks to the server using LSP
- Letting the IDE directly import the Cursorless engine as a Javascript NPM package. If the IDE doesn't natively support JS we can use https://bellard.org/quickjs/
- Bundle Cursorless into a wasm and let the IDE run the Cursorless wasm

In all cases, the Cursorless core maintainers will be responsible for the part of the Cursorless inside the “interface boundary”, ie the Cursorless LSP server, the Cursorless NPM package, or the Cursorless wasm. The IDE-specific code that interacts with the Cursorless interface will be maintained by someone who is responsible specifically for the integration with the given IDE.

Given that we are considering three possible ways to integrate, we have decided not to go with a waterfall approach where we invest in picking one approach, implementing it, then letting IDE-specific contributor build the client.

Instead, we’d like to work closely with a couple IDE client maintainers so that we can figure out which approach makes the most sense for their IDE. If you’re interested in being one of those IDE client maintainers, and have some capacity to hack in the next couple months; leave a message in this discussion, and then drop into a meet-up so we can get started!

See the repos in https://github.com/cursorless-everywhere/ for some of the current work

**Comment by pokey** [comment]
JetBrains

**Reply by pokey** [reply]
@phillco is currently leading the charge here

**Reply by pokey** [reply]
I believe the relevant repos are:

- https://github.com/cursorless-everywhere/cursorless
- https://github.com/cursorless-everywhere/cursorless-jetbrains
- https://github.com/cursorless-everywhere/sidecar

but @phillco please correct me if I'm wrong

**Reply by Josh-Piper** [reply]
Thanks so much @phillco . Cant wait to see it, happy to be an alpha/beta tester

**Reply by jaresty** [reply]
Me too! Sign me up for the beta!

**Reply by pokey** [reply]
@AndreasArvidsson and @phillco are considering doing this using https://bellard.org/quickjs/

One thought: do we want to use in-memory text document or thin wrapper that always reads source of truth from jetbrains? The latter saves needing to synchronise

**Reply by pokey** [reply]
Also fwiw I believe @johneffo / @johnefford got a bunch of the cursorless recorded tests running in jetbrains. Might be worth seeing if you can borrow that code

**Comment by pokey** [comment]
emacs

**Reply by pokey** [reply]
@rntz and @dyamito have done most of the work in this direction

**Reply by pokey** [reply]
See:

- https://github.com/cursorless-everywhere/sidecar
- https://github.com/cursorless-everywhere/emacs
- https://github.com/cursorless-everywhere/emacs-command-server
- https://github.com/cursorless-everywhere/emacs-cursorless

**Comment by pokey** [comment]
vim / neovim / terminal

**Reply by pokey** [reply]
Not sure how much has been done here

**Reply by pokey** [reply]
see https://github.com/neovim/neovim/issues/21603 for discussion of how to draw the hats

**Reply by pokey** [reply]
Just an update that @saidelike is starting to look into Cursorless for neovim, and https://neovim.io/node-client/ is looking like a promising approach. cc/ @fidgetingbits

**Reply by saidelike** [reply]
Progress tracked in https://github.com/cursorless-dev/cursorless/pull/2256

**Comment by pokey** [comment]
visual studio

**Comment by conartist6** [comment]
BABLR

**Reply by conartist6** [reply]
It wouldn't be the world's most accessible editing code editing interface without Cursorless so I'll lead the charge there, but I have to finish writing the platform before I can write the Cursorless integration.

**Comment by pokey** [comment]
RStudio

**Comment by pokey** [comment]
Godot

**Reply by pokey** [reply]
Some links:

- https://github.com/nonunknown/godot-powerful
- https://docs.godotengine.org/en/stable/classes/class_editorplugin.html
- https://docs.godotengine.org/en/stable/classes/class_editorinterface.html#class-editorinterface

**Reply by awulkan** [reply]
Rendering in code editor:
https://github.com/jotson/ridiculous_coding

**Reply by rinOfTheStars** [reply]
This obviously will need GDScript support in Cursorless itself before it can really begin. On the subject of integrating Cursorless into the engine and editor, https://github.com/ashtonmeuser/godot-wasm exists and so WASM is likely the best option—with the voice-to-input accessibility plugin that would provide Cursorless integration depending on the godot-wasm plugin.

I'm more than willing to give working on such a Godot plugin a shot once someone more knowledgeable when it comes to adding new languages to Cursorless than I am gets around to GDScript support.

**Comment by saidelike** [comment]
Ghidra (reverse engineering app)

**Reply by saidelike** [reply]
Some links

- https://ghidra-sre.org/
- https://github.com/vaguue/Ghidra.js
- https://github.com/er28-0652/setup-ghidra
